### PR TITLE
Rtc service

### PR DESCRIPTION
**Pull request by LePailleurThibault**
Implementation of the RTC service sending periodically time to the network.

### CODE REVIEW

File: docker/rtc_service/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,63 @@
+FROM python:3.8.8-alpine3.12 AS builder
```
Comment (GwendalRaoul) [inline docker/rtc_service/Dockerfile:1]: It has to be changed to 3.17 now as it is the base image for other services

File: docker/rtc_service/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+USER wirepas
+WORKDIR /home/wirepas
+
+COPY --chown=wirepas ./python_transport /home/wirepas/python_transport
+WORKDIR /home/wirepas/python_transport
+
+RUN ./utils/generate_wheel.sh
+
+USER wirepas
+
```
Comment (GwendalRaoul) [inline docker/rtc_service/Dockerfile:28]: Add a note here that this is a hack as dbus client is part of python_transport, but later they should be splitted

File: rtc_service/rtc_service.py [hunk 3]
```diff
@@ -0,0 +1,288 @@
...
+        self.synchronization_thread.start()
+
+    def on_stack_started(self, name):
+        logging.debug("Sink started: %s", name)
+
+    def on_stack_stopped(self, name):
+        logging.debug("Sink stopped: %s", name)
+
+    def on_sink_connected(self, name):
+        logging.info("Sink connected, sending new configs")
```
Comment (GwendalRaoul) [inline rtc_service/rtc_service.py:183]: Is the message true ?
You say, "sending new configs" but nothing is sent!

File: rtc_service/rtc_service.py [hunk 4]
```diff
@@ -0,0 +1,288 @@
...
+        self.synchronization_thread = SynchronizationThread(
+            settings.rtc_synchronization_period_s,
+            settings.timezone_offset_s,
+            settings.timezone_from_gateway_clock,
+            settings.get_time_from_local,
+            self.sink_manager,
+        )
+        self.synchronization_thread.start()
+
+    def on_stack_started(self, name):
```
Comment (GwendalRaoul) [inline rtc_service/rtc_service.py:176]: Is it really needed to override those calls for just a debug message?

File: rtc_service/rtc_service.py [hunk 5]
```diff
@@ -0,0 +1,288 @@
...
+        Main service for rtc service module
+    """
+    parser = argparse.ArgumentParser(fromfile_prefix_chars='@')
+
+    parser.add_argument(
+        "--rtc_synchronization_period_s",
+        default=os.environ.get("WM_RTC_SYNCHRONIZATION_PERIOD_S", 20*60),
+        action="store",
+        type=str2int,
+        help=("Period of time before sending a new rtc time in the network."),
```
Comment (GwendalRaoul) [inline rtc_service/rtc_service.py:224]: before sending -> to send

File: rtc_service/rtc_service.py [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+              "True means that the time is taken directly from gateway."
+              "Note: You must assure that gateway are synchronized if set to True"),
+    )
+
+    settings = parser.parse_args()
+
+    # Set default debug level
+    debug_level = "info"
+    try:
+        debug_level = os.environ["DEBUG_LEVEL"]
```
Comment (GwendalRaoul) [inline rtc_service/rtc_service.py:261]: This is a new service so you don't have to keep backward compatibility

File: rtc_service/rtc_service.py [hunk 7]
```diff
@@ -0,0 +1,288 @@
...
+        """
+        Main loop that send periodically gateway time to the network.
+        """
+        self.running = True
+
+        while self.running:
+            self.publish_time()
+            sleep(self.period)
+
+    def stop(self):
```
Comment (GwendalRaoul) [inline rtc_service/rtc_service.py:152]: It looks like you never call stop.
Is it needed to keep it?

Thread is daemonized so should be killed when the process is closed

**Review by GwendalRaoul — CHANGES_REQUESTED** [review]

File: rtc_service/rtc_service.py (hunk 7 above)
Comment (LePailleurThibault) [reply rtc_service/rtc_service.py:152]: I'm going to remove this function. 
stop() function in the transport service is also never called and it's also daemonized. Should we remove this one as well ?

File: rtc_service/rtc_service.py (hunk 4 above)
Comment (LePailleurThibault) [reply rtc_service/rtc_service.py:176]: Logs are useful but, you are right. If the stack starts, at least the transport service should logs this message. We don't need to duplicate the logs.

File: rtc_service/rtc_service.py [hunk 8]
```diff
@@ -0,0 +1,263 @@
...
+                hop_limit=0,
+            )
+            logging.debug("time elapsed to send RTC time through the gateway: "
+                          f"{int((time()-start)*1000)}ms")
+
+    def run(self):
+        """
+        Main loop that send periodically gateway time to the network.
+        """
+        self.running = True
```
Comment (GwendalRaoul) [inline rtc_service/rtc_service.py:147]: self.running is useless

File: rtc_service/rtc_service.py [hunk 9]
```diff
@@ -0,0 +1,263 @@
...
+        data_payload = RTC_VERSION.to_bytes(2, "little") + encode_tlv(timestamp_ms, self.timezone_offset_s)
+        logging.info("Send rtc message to the network")
+        for sink in self.sink_manager.get_sinks():
+            start = time()
+            sink.send_data(
+                dst=BROADCAST_ADDRESS,
+                src_ep=RTC_SOURCE_EP,
+                dst_ep=RTC_DEST_EP,
+                qos=0,
+                initial_time=INITIAL_DELAY_MS,
```
Comment (GwendalRaoul) [inline rtc_service/rtc_service.py:135]: you can remove this line, it is deprecated feature

**Review by GwendalRaoul — APPROVED** [review]

**Pull request merged**